### PR TITLE
Drnet performance updates

### PR DESCRIPTION
**Pull request by greggcz**
Fixes SQL port update (deleted=>1, port_id = '') which should not happen
Introduces dbBulkInsertUpdate to optimize SQL queries (devices with large amount of ports & using remote database with higher latency suffer from SQL roundtrip issues resulting in never ending discover and polls) see ports.inc.php and dbFacile.mysql.inc.php

-Gregg

### CODE REVIEW

File: includes/dbFacile.mysql.php [hunk 1]
```diff
@@ -147,13 +208,13 @@ function dbBulkInsert($data, $table) {
...
         }
-        $values .= "(".$rowvalues.")";
+        $values .= "(".$rowvalues.")\n";
     }
 
     $time_start = microtime(true);
     $result = dbQuery($sql.$values);
 
-    // logfile($fullSql);
+    logfile($fullSql);
```
Comment (laf) [inline includes/dbFacile.mysql.php]: This should remain commented out.

File: includes/dbFacile.mysqli.php [hunk 2]
```diff
@@ -153,7 +153,7 @@ function dbBulkInsert($data, $table) {
     $time_start = microtime(true);
     $result = dbQuery($sql.$values);
 
-    // logfile($fullSql);
+    logfile($fullSql);
```
Comment (laf) [inline includes/dbFacile.mysqli.php]: This should remain commented out.

File: includes/discovery/entity-physical.inc.php [hunk 3]
```diff
@@ -124,15 +132,54 @@
 
                 if (!empty($ifIndex)) {
                     $insert_data['ifIndex'] = $ifIndex;
-                }
+                } else {
```
Comment (laf) [inline includes/discovery/entity-physical.inc.php]: This should be:

}
else {

File: includes/discovery/entity-physical.inc.php [hunk 4]
```diff
@@ -124,15 +132,54 @@
...
+		    $insert_data['ifIndex'] = null;
+		}
 
-                dbInsert($insert_data, 'entPhysical');
+                array_push($delayed_insert, $insert_data);
+                //dbInsert($insert_data, 'entPhysical');
```
Comment (laf) [inline includes/discovery/entity-physical.inc.php]: Remove this line. git has a history :)

**Comment by laf** [comment]
I've commented inline on a few things but in general:

} else {

should always be:

}
else {

(Check our code guideline docs for info on our set standards).

This seems like it will save a good number of queries so I'm all for this. dbBulkInsertUpdate() will need to be added to mysqli dbFacile as well.

What's the reason to add -Cr50 to all snmp queries?

**Comment by greggcz** [comment]
Still working on few changes using the new dbBulkInsertUpdate() and will add it to mysqli as well.
So far with mysql latency of about 105ms and device latency of 25ms, my changes reduced time to poll device with hundreds of ports and vlans from 160s down to 90s.

Regarding -Cr 50 for 'snmpbulkwalk', it increases max repetitions for snmp:

"-Cr<NUM>
Set the max-repetitions field in the GETBULK PDUs. This specifies the maximum number of iterations over the repeating variables. The default is 10."

Increasing it to 50 allows the UDP packet to be larger therefore requires less packets per second and increases performance. IMO it should be a setting in the config or even per device but from tests it speeds up snmpbulkwalk queries by 15 to 30% depending on latency to the device.

-Gregg

**Comment by laf** [comment]
Thanks @greggcz.

When you've got a final version let me know and I'd be happy to test on a 2800+ device install :)

**Comment by laf** [comment]
@greggcz Any updates?

**Comment by greggcz** [comment]
Testing changes now, will merge into the branch early next week

File: includes/dbFacile.mysql.php [hunk 5]
```diff
@@ -107,6 +107,67 @@ function dbInsert($data, $table) {
...
+    }
+    foreach($keys as $key) {
+	$update_keys .= "`$key`=VALUES(`$key`),";
+    }
+    $update_keys = ' ON DUPLICATE KEY UPDATE ' . trim($update_keys, ',') . ' ';
+
+    $time_start = microtime(true);
+    $result = dbQuery($sql.$values.$update_keys);
+
+    logfile($fullSql);
```
Comment (paulgear) [inline includes/dbFacile.mysql.php]: If this is going to be used for bulk data, we probably should comment this out as well, or at least only log the first line or so.

File: includes/discovery/ipv4-addresses.inc.php [hunk 6]
```diff
@@ -4,16 +4,32 @@
...
+$port_db_ids = dbFetchRows('SELECT `ifIndex`,`port_id` FROM `ports` WHERE `device_id` = ?',array($device['device_id']));
+$port_ids = array();
+foreach ($port_db_ids as $entry) {
+	$port_ids[$entry['ifIndex']] = $entry['port_id'];
+}
 foreach (explode("\n", $oids) as $data) {
     $data               = trim($data);
     list($oid,$ifIndex) = explode(' ', $data);
-    $mask               = trim(snmp_get($device, "ipAdEntNetMask.$oid", '-Oqv', 'IP-MIB'));
+    //$mask               = trim(snmp_get($device, "ipAdEntNetMask.$oid", '-Oqv', 'IP-MIB'));
```
Comment (paulgear) [inline includes/discovery/ipv4-addresses.inc.php]: More commented out lines here; if they're not needed, just delete them.

File: includes/discovery/vlans/cisco-vtp.inc.php [hunk 7]
```diff
@@ -12,17 +12,21 @@
...
+	    $db_to_insert = array();
             foreach ($vlans[$vtpdomain_id] as $vlan_id => $vlan) {
                 echo " $vlan_id";
                 if (is_array($vlans_db[$vtpdomain_id][$vlan_id])) {
                     echo '.';
                 }
                 else {
-                    dbInsert(array('device_id' => $device['device_id'], 'vlan_domain' => $vtpdomain_id, 'vlan_vlan' => $vlan_id, 'vlan_name' => $vlan['vtpVlanName'], 'vlan_type' => $vlan['vtpVlanType']), 'vlans');
+		    array_push($db_to_insert(array('device_id' => $device['device_id'], 'vlan_domain' => $vtpdomain_id, 'vlan_vlan' => $vlan_id, 'vlan_name' => $vlan['vtpVlanName'], 'vlan_type' => $vlan['vtpVlanType'])));
+                    //dbInsert(array('device_id' => $device['device_id'], 'vlan_domain' => $vtpdomain_id, 'vlan_vlan' => $vlan_id, 'vlan_name' => $vlan['vtpVlanName'], 'vlan_type' => $vlan['vtpVlanType']), 'vlans');
```
Comment (paulgear) [inline includes/discovery/vlans/cisco-vtp.inc.php]: Delete

File: includes/polling/ports.inc.php [hunk 8]
```diff
@@ -558,8 +559,10 @@
 
         // Update Database
         if (count($port['update'])) {
-            $updated = dbUpdate($port['update'], 'ports', '`port_id` = ?', array($port['port_id']));
-            d_echo("$updated updated");
+            //$updated = dbUpdate($port['update'], 'ports', '`port_id` = ?', array($port['port_id']));
```
Comment (paulgear) [inline includes/polling/ports.inc.php]: Delete

**Comment by paulgear** [comment]
It will be interesting to see what effect the -Cr50 has on general polling performance.  Do we know of anything which it might break?  It seems to me that if it doesn't break anything we should consider increasing it even more, say to something in the 100-1000 range, since that's closer to the likely maximum number of ports on many large switches.

**Comment by SaaldjorMike** [comment]
Any updates on this one?

**Comment by laf** [comment]
Closing for now. @greggcz please feel free to re-open this or create a new issue.

**Pull request closed without merging**